pkg/network/encoding/format.go

```diff
@@ -45,78 +39,109 @@ func (ipc ipCache) Get(addr util.Address) string {
 }
 
 // FormatConnection converts a ConnectionStats into an model.Connection
-func FormatConnection(
-	conn network.ConnectionStats,
-	routes map[string]RouteIdx,
-	httpEncoder *httpEncoder,
-	http2Encoder *http2Encoder,
-	kafkaEncoder *kafkaEncoder,
-	dnsFormatter *dnsFormatter,
-	ipc ipCache,
-	tagsSet *network.TagsSet,
-) *model.Connection {
-	c := connPool.Get().(*model.Connection)
-	c.Pid = int32(conn.Pid)
+func FormatConnection(builder *model.ConnectionBuilder, conn network.ConnectionStats, routes map[string]RouteIdx, httpEncoder *httpEncoder, http2Encoder *http2Encoder, kafkaEncoder *kafkaEncoder, dnsFormatter *dnsFormatter, ipc ipCache, tagsSet *network.TagsSet) {
+
+	builder.SetPid(int32(conn.Pid))
+
 	var containerID string
 	if conn.ContainerID != nil {
 		containerID = *conn.ContainerID
 	}
-	c.Laddr = formatAddr(conn.Source, conn.SPort, containerID, ipc)
-	c.Raddr = formatAddr(conn.Dest, conn.DPort, "", ipc)
-	c.Family = formatFamily(conn.Family)
-	c.Type = formatType(conn.Type)
-	c.IsLocalPortEphemeral = formatEphemeralType(conn.SPortIsEphemeral)
-	c.LastBytesSent = conn.Last.SentBytes
-	c.LastBytesReceived = conn.Last.RecvBytes
-	c.LastPacketsSent = conn.Last.SentPackets
-	c.LastPacketsReceived = conn.Last.RecvPackets
-	c.LastRetransmits = conn.Last.Retransmits
-	c.Direction = formatDirection(conn.Direction)
-	c.NetNS = conn.NetNS
-	c.RemoteNetworkId = ""
-	c.IpTranslation = formatIPTranslation(conn.IPTranslation, ipc)
-	c.Rtt = conn.RTT
-	c.RttVar = conn.RTTVar
-	c.IntraHost = conn.IntraHost
-	c.LastTcpEstablished = conn.Last.TCPEstablished
-	c.LastTcpClosed = conn.Last.TCPClosed
-	c.Protocol = formatProtocolStack(conn.ProtocolStack, conn.StaticTags)
-
-	c.RouteIdx = formatRouteIdx(conn.Via, routes)
-	dnsFormatter.FormatConnectionDNS(conn, c)
+
+	builder.SetLaddr(func(w *model.AddrBuilder) {
+		w.SetIp(ipc.Get(conn.Source))
+		w.SetPort(int32(conn.SPort))
+		w.SetContainerId(containerID)
+	})
+	builder.SetRaddr(func(w *model.AddrBuilder) {
+		w.SetIp(ipc.Get(conn.Dest))
+		w.SetPort(int32(conn.DPort))
+	})
```

Comment:
you must check conn.Source and conn.Dest are not nil, otherwise `ipc.Get` will panic

Reply:
Are there instances where source/dest are nil?  IMO those should be filtered out before we get to encoding.

This worked fine on the testing environment, and passes unit tests, so if it's a possibility that's a pointer to us missing some tests.

Reply:
the original formatAddr checked it, I'm not against of filtering it out, I'm only concern that the code "was there for a reason"

Reply:
[seeing](https://github.com/DataDog/datadog-agent/pull/3783) @sunhay as the original author 4 years ago

Reply:
Thanks for hunting that down. 

Given the age of the code, and that this ran on several internal clusters already, I'll leave the code as is. Let me know if you disagree.